python/ray/serve/_private/http_util.py

```diff
@@ -456,3 +457,37 @@ def validate_http_proxy_callback_return(
                     f"instead got {type(middleware)} type item in the list."
                 )
     return middlewares
+
+
+class GaugeSet:
+    """Set that tracks its size using a Prometheus gauge.
+
+    The HTTPProxyActor's uvicorn server stores each connection in a set.
+    The HTTPProxyActor replaces that set with an instance of this class to
+    track the number of connections.
+
+    This class keeps items in an internal set, and it updates a Prometheus
+    gauge with the the set's size whenever the `add`, `remove`, or `discard`
+    methods are called.
+    """
+
+    def __init__(self, connection_gauge: metrics.Gauge):
+        self.connection_gauge = connection_gauge
+        self.internal_set = set()
+
+    def add(self, element: Any):
```

Comment:
Non-blocker: the element should probably be a protocol type https://github.com/encode/uvicorn/blob/master/uvicorn/server.py#L44